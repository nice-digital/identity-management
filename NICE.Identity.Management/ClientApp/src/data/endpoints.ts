const API_BASE_URL = process.env.REACT_APP_API_BASE_URL
	? process.env.REACT_APP_API_BASE_URL
	: "#{REACT_APP_API_BASE_URL}";

export const Endpoints = {
	usersList: `${API_BASE_URL}/users`, // get all users
	user: (userId: string | number): string => `${API_BASE_URL}/users/${userId}`,

	servicesList: `${API_BASE_URL}/services`,
	service: (serviceId: string | number): string =>
		`${API_BASE_URL}/services/${serviceId}`,
	
	userRolesByWebsite: (userId: string | number, websiteId: string | number): string =>
		`${API_BASE_URL}/users/${userId}/rolesbywebsite/${websiteId}`,

	verificationEmail: `${API_BASE_URL}/VerificationEmail/VerificationEmail`,
	websitesList: `${API_BASE_URL}/websites`,
};
