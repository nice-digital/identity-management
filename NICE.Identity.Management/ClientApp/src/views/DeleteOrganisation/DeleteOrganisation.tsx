import React, { Component } from "react";
import { RouteComponentProps, Link } from "react-router-dom";

import { Alert } from "@nice-digital/nds-alert";
import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { Button } from "@nice-digital/nds-button";
import { PageHeader } from "@nice-digital/nds-page-header";
import { Grid, GridItem } from "@nice-digital/nds-grid";

import { DeleteOrganisationConfirmation } from "../DeleteOrganisationConfirmation/DeleteOrganisationConfirmation";
import { OrganisationType } from "../../models/types";
import { Endpoints } from "../../data/endpoints";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";

import { fetchData } from "../../helpers/fetchData";
import { isDataError } from "../../helpers/isDataError";

type TParams = { id: string };

type DeleteOrganisationProps = Record<string, unknown> &
	RouteComponentProps<TParams>;

type DeleteOrganisationState = {
	organisation: OrganisationType;
	hasBeenDeleted: boolean;
	error?: Error;
	isLoading: boolean;
	isDeleteButtonLoading: boolean;
};

export class DeleteOrganisation extends Component<
	DeleteOrganisationProps,
	DeleteOrganisationState
> {
	constructor(props: DeleteOrganisationProps) {
		super(props);

		this.state = {
			organisation: {} as OrganisationType,
			hasBeenDeleted: false,
			isLoading: true,
			isDeleteButtonLoading: false,
		};
	}

	handleDeleteClick = async (): Promise<void> => {
		this.setState({ isDeleteButtonLoading: true });

		const fetchOptions = {
			method: "DELETE",
			headers: { "Content-Type": "application/json" },
			body: JSON.stringify({ id: this.props.match.params.id }),
		};

		const deletedOrganisation = await fetchData(
			Endpoints.organisation(this.props.match.params.id),
			fetchOptions,
		);

		if (isDataError(deletedOrganisation)) {
			this.setState({ error: deletedOrganisation });
		} else {
			this.setState({ hasBeenDeleted: true });
		}

		this.setState({ isDeleteButtonLoading: false });
	};

	async componentDidMount(): Promise<void> {
		this.setState({ isLoading: true });

		const organisation = await fetchData(
			Endpoints.organisation(this.props.match.params.id),
		);

		if (isDataError(organisation)) {
			this.setState({ error: organisation });
		}

		this.setState({ organisation, isLoading: false });
		document.title = `NICE Accounts - Delete ${organisation.name}`;
	}

	render(): JSX.Element {
		const {
			isLoading,
			isDeleteButtonLoading,
			error,
			hasBeenDeleted,
			organisation,
		} = this.state;
		const { id } = this.props.match.params;

		let lastBreadcrumb = organisation.name;

		if (isLoading) {
			lastBreadcrumb = "Loading organisation details";
		}

		return (
			<>
				{hasBeenDeleted ? (
					<DeleteOrganisationConfirmation name={organisation.name} />
				) : (
					<>
						<Breadcrumbs>
							<Breadcrumb
								data-qa-sel="breadcrumb-administration-link"
								to="/"
								elementType={Link}
							>
								Administration
							</Breadcrumb>
							<Breadcrumb
								data-qa-sel="breadcrumb-organisation-link"
								to="/organisations"
								elementType={Link}
							>
								Organisations
							</Breadcrumb>
							{error ? (
								<Breadcrumb>Error</Breadcrumb>
							) : (
								<Breadcrumb to={`/organisations/${id}`} elementType={Link}>
									{lastBreadcrumb}
								</Breadcrumb>
							)}
							<Breadcrumb>Delete organisation</Breadcrumb>
						</Breadcrumbs>

						{!error ? (
							<>
								{isLoading ? (
									<>
										<PageHeader
											preheading="Confirm"
											heading="Delete organisation"
										/>
										<Grid>
											<GridItem cols={12} md={9}>
												<p>Loading...</p>
											</GridItem>
										</Grid>
									</>
								) : (
									<>
										<PageHeader heading={`${organisation.name}`} />
										<Grid>
											<GridItem cols={12} md={9}>
												<Alert
													type="info"
													role="status"
													aria-live="polite"
													data-qa-sel="warning-message-delete-organisation"
												>
													<p>
														Once you delete an organisation, the action cannot
														be undone. Are you sure you want to proceed?
													</p>
												</Alert>

												<Button
													data-qa-sel="confirm-delete-organisation"
													variant="cta"
													onClick={this.handleDeleteClick}
													disabled={isDeleteButtonLoading}
												>
													{isDeleteButtonLoading ? "Loading..." : "Confirm"}
												</Button>
												<Button
													to={`/organisations/${id}`}
													variant="secondary"
													elementType={Link}
													disabled={isDeleteButtonLoading}
												>
													Cancel
												</Button>
											</GridItem>
										</Grid>
									</>
								)}
							</>
						) : (
							<>
								<PageHeader heading="Error" />
								<ErrorMessage error={error}></ErrorMessage>
							</>
						)}
					</>
				)}
			</>
		);
	}
}
