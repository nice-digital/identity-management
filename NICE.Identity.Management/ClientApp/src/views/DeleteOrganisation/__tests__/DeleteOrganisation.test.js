import React from "react";
import { render, waitForElementToBeRemoved, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { rest } from "msw";
import { setupServer } from "msw/node";
import { MemoryRouter } from "react-router";
import { DeleteOrganisation } from "../DeleteOrganisation";
import { Endpoints } from "src/data/endpoints";
import singleOrganisation from "./singleOrganisation.json";

const consoleErrorReset = console.error;

beforeEach(() => {
	console.error = consoleErrorReset;
});

const server = setupServer(
  rest.get(Endpoints.organisation("1"), (req, res, ctx) => {
    return res(ctx.json(singleOrganisation));
  }),
  rest.delete(Endpoints.organisation("1"), (req, res, ctx) => {
    return res(ctx.json({}));
  })
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

test("should show loading message before data has been loaded", () => {
  const match = {
    params: { id: 1 },
    isExact: true,
    path: "",
    url: "",
  };
  render(<DeleteOrganisation match={match} />, {wrapper: MemoryRouter});
  expect(screen.getByText("Loading...")).toBeInTheDocument();
});

test("should match the snapshot after data has been loaded", async () => {
	const match = {
		params: { id: 1 },
		isExact: true,
		path: "",
		url: "",
	};
	const {container} = render(<DeleteOrganisation match={match} />, {wrapper: MemoryRouter});
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	expect(container).toMatchSnapshot();
});

test("should show error message when fetchData function returns 401 error", async () => {
	console.error = jest.fn();
	const match = {
		params: { id: 1 },
		isExact: true,
		path: "",
		url: "",
	};
	server.use(
		rest.get(Endpoints.organisation(match.params.id), (req, res, ctx) => {
			return res.once(
				ctx.status(401),
				ctx.json({})
			);
		})
	);
	const {container} = render(<DeleteOrganisation match={match} />, {wrapper: MemoryRouter});
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	expect(container).toMatchSnapshot();
});

test("should show error message when fetchData function returns 500 error", async () => {
	console.error = jest.fn();
	const match = {
		params: { id: 1 },
		isExact: true,
		path: "",
		url: "",
	};
	server.use(
		rest.get(Endpoints.organisation(match.params.id), (req, res, ctx) => {
			return res.once(
				ctx.status(500),
				ctx.json({})
			);
		})
	);
	const {container} = render(<DeleteOrganisation match={match} />, {wrapper: MemoryRouter});
	await waitForElementToBeRemoved(() => screen.getByText("Loading...", { selector: "p" }));
	expect(container).toMatchSnapshot();
});

test("should show error message when fetchData delete fails", async () => {
	console.error = jest.fn();
	const match = {
		params: { id: 1 },
		isExact: true,
		path: "",
		url: "",
	};
	server.use(
		rest.get(Endpoints.organisation(match.params.id), (req, res, ctx) => {
			return res.once(
				ctx.json({})
			);
		}),
		rest.delete(Endpoints.organisation(match.params.id), (req, res, ctx) => {
			return res.once(
				ctx.status(500),
				ctx.json({})
			);
		})
	);
	render(<DeleteOrganisation match={match} />, {wrapper: MemoryRouter});
	const confirmButton = await screen.findByText("Confirm", { selector: "button" });
	userEvent.click(confirmButton);
	const errorMessage = await screen.findByRole("heading", { name: "Error" });
	expect(errorMessage).toBeInTheDocument();
});

test("should show error message when fetchData delete returns non-200 error", async () => {
	console.error = jest.fn();
	const match = {
		params: { id: 1 },
		isExact: true,
		path: "",
		url: "",
	};
	server.use(
		rest.get(Endpoints.organisation(match.params.id), (req, res, ctx) => {
			return res.once(
				ctx.json({})
			);
		}),
		rest.delete(Endpoints.organisation(match.params.id), (req, res, ctx) => {
			return res.once(
				ctx.status(401),
				ctx.json({})
			);
		})
	);
	render(<DeleteOrganisation match={match} />, {wrapper: MemoryRouter});
	const confirmButton = await screen.findByText("Confirm", { selector: "button" });
	userEvent.click(confirmButton);
	const errorMessage = await screen.findByRole("heading", { name: "Error" });
	expect(errorMessage).toBeInTheDocument();
});

test("should disable delete button when clicked", async () => {
	const match = {
		params: { id: 1 },
		isExact: true,
		path: "",
		url: "",
	};
	render(<DeleteOrganisation match={match} />, {wrapper: MemoryRouter});
	let confirmButton = await screen.findByText("Confirm", { selector: "button" });
	userEvent.click(confirmButton);
	confirmButton = await screen.findByRole("button", { name: "Loading..." });
	expect(confirmButton).toHaveAttribute("disabled");
	expect(confirmButton).toBeInTheDocument();
});

test("should display confirmation message once fetchData delete is successfully complete", async () => {
	const match = {
		params: { id: 1 },
		isExact: true,
		path: "",
		url: "",
	};
	render(<DeleteOrganisation match={match} />, {wrapper: MemoryRouter});
	const confirmButton = await screen.findByText("Confirm", { selector: "button" });
	userEvent.click(confirmButton);
	const successMessage = await screen.findByRole("heading", { name: "Organisation deleted" });
	expect(successMessage).toBeInTheDocument();
});