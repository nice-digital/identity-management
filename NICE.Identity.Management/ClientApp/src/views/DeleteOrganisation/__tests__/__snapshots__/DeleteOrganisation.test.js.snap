// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot after data has been loaded 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-organisation-link"
          href="/organisations"
        >
          Organisations
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/organisations/1"
        >
          NICE
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Delete organisation
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/organisations",
        "name": "Organisations"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/organisations/1",
        "name": "NICE"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "name": "Delete organisation"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading"
    >
      NICE
    </h1>
  </div>
  <div
    class="grid"
  >
    <div
      data-g="12 md:9"
    >
      <div
        aria-live="polite"
        class="alert alert--info"
        data-qa-sel="warning-message-delete-organisation"
        role="status"
      >
        <p>
          Once you delete an organisation, the action cannot be undone. Are you sure you want to proceed?
        </p>
      </div>
      <button
        class="btn btn--cta"
        data-qa-sel="confirm-delete-organisation"
        type="button"
      >
        Confirm
      </button>
      <a
        class="btn btn--secondary"
        href="/organisations/1"
      >
        Cancel
      </a>
    </div>
  </div>
</div>
`;

exports[`should show error message when fetchData function returns 401 error 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-organisation-link"
          href="/organisations"
        >
          Organisations
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Error
        </span>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Delete organisation
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/organisations",
        "name": "Organisations"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "name": "Error"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "name": "Delete organisation"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading"
    >
      Error
    </h1>
  </div>
  <div
    class="alert alert--error"
    role="alert"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>
</div>
`;

exports[`should show error message when fetchData function returns 500 error 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-organisation-link"
          href="/organisations"
        >
          Organisations
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Error
        </span>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Delete organisation
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/organisations",
        "name": "Organisations"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "name": "Error"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "name": "Delete organisation"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading"
    >
      Error
    </h1>
  </div>
  <div
    class="alert alert--error"
    role="alert"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>
</div>
`;
